src/main/java/com/google/sps/servlets/AddSaleCardServlet.java

```diff
@@ -0,0 +1,173 @@
+        vendorLocation = new GeoPt(latitude, longitude);
+      } catch (DateTimeParseException e) {
+        System.out.println("Bad format to parse: " + e);
+        response.sendError(HttpServletResponse.SC_BAD_REQUEST);
+        return;
+      } catch (NumberFormatException e) {
+        System.out.println("The string is not a parsable float: " + e);
+        response.sendError(HttpServletResponse.SC_BAD_REQUEST);
+        return;
+      } catch (IllegalArgumentException e) {
+        System.out.println("Latitude and/or longitude outside legal range: " + e);
+        response.sendError(HttpServletResponse.SC_BAD_REQUEST);
+        return;
+      }
+
+      // Get the BlobKey of the image that the user uploaded to Blobstore
+      BlobKey imageBlobKey = getUploadedFileBlobKey(request, "imageFile");
+      String altText = HttpServletUtils.getParameter(request, "altText", "");
+
+      // Check values are not empty or outside range
+      if (businessName.isEmpty() || description.isEmpty() || geoHash.isEmpty() || altText.isEmpty() || 
+          imageBlobKey == null || radius < MIN_DISTANCE || radius > MAX_DISTANCE ) {
+        response.sendError(HttpServletResponse.SC_BAD_REQUEST);
+        return;
+      }
+
+      DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
+
+      // TODO: Create vendor key using the Auth ID
+      // Key vendorKey = KeyFactory.createKey("Vendor", (String) user.authID);
+
+      // Generate Key when datastore put
+      Entity saleCard = new Entity("SaleCard"/*, vendorKey*/);
+      datastore.put(saleCard);      
+
+      // TODO: Remove from datastore existing saleCard, picture and location data
+
+      Entity picture = new Entity("Picture", saleCard.getKey());
+      picture.setProperty("blobKey", imageBlobKey);
+      picture.setProperty("altText", altText);
+      datastore.put(picture);
+
+      Entity locationData = new Entity("LocationData", saleCard.getKey());
+      locationData.setProperty("salePoint", vendorLocation);
+      locationData.setProperty("geoHash", geoHash);
+      locationData.setProperty("radius", radius);
+      datastore.put(locationData);
+
+      EmbeddedEntity picInfo = new EmbeddedEntity();
+      picInfo.setPropertiesFrom(picture);
+      EmbeddedEntity locInfo = new EmbeddedEntity();
+      locInfo.setPropertiesFrom(locationData);
+
+      saleCard.setIndexedProperty("picture", picInfo);
+      saleCard.setIndexedProperty("location", locInfo);
+      datastore.put(saleCard);
+
+      EmbeddedEntity saleInfo = new EmbeddedEntity();
+      saleInfo.setPropertiesFrom(saleCard);
+
+      /* TODO: Vendor Entity Put
+        Entity vendor = datastore.get(vendorKey);
+        vendor.setIndexedProperty("saleCard", saleInfo);
+        datastore.put(vendor);
+      */
+
+      response.sendRedirect("/");
+    } else {
+      response.sendError(HttpServletResponse.SC_UNAUTHORIZED);
```

Comment:
Question: When fetching to this Servlet on the front end, what exactly will `response.sendError(...) `send? Is it a string with an error message, boolean maybe?

Reply:
Oh, SC stands for Status Code, in this case is the code for Unauthorized, which is 401, so it is an integer. In the front end, you can check the status code like this `response.status`.
In case you want to [know more about other status codes.](https://developer.mozilla.org/en-US/docs/Web/HTTP/Status) :)